<img src="https://marketing4ecommerce.net/wp-content/uploads/2015/09/logo-iebs.jpg" style="float:right" width="400">

# Introducción a los lenguajes de programación

## Actividad Semanal de la clase 2

### Javier Cózar

*Entregado por Bruno Anglés el 5/12/2022*


# Dataset


En esta libreta vamos a trabajar con un conjunto de datos que contiene los datos de la FIFA de la copa del mundo. Disponéis de los archivos ya descargados en la sección de recursos. Utilizaremos estos tres ficheros:

- WorldCupMatches.csv: Datos de los partidos disputados
- WorldCupPlayers.csv: Datos de los jugadores que han jugado en cada partido
- WorldCups.csv: Datos de las copas disputadas y los resultados

# Introducción

Mediante esta libreta vamos a explorar los datos almacenados en estos tres ficheros y a responder preguntas haciendo operaciones y transformaciones con pandas. Además de completar las celdas de código para responder a los ejercicios, cuando así se indique se deberá añadir una celda de tipo _markdown_ explicando con texto enriquecido la información o las conclusiones extraídas.

In [1]:
import pandas as pd

In [59]:
# Cargamos el conjunto de datos WorldCupMatches
df_partidos = pd.read_csv("WorldCupMatches.csv", parse_dates=["Datetime"])

## 1. Exploración

Usar las funciones `head`, `info` y `describe` para explorar el dataframe `df_partidos`. **Crear una celda de tipo markdown** y comentar brevemente la información almacenada en cada dataframe (columnas y tipo de datos).

_No es necesario hacer una documentación exhaustiva, tan solo comentar aspectos principales como qué tipo de información almacena el DataFrame, cuántas filas y columnas tiene, y comentar las columnas que se consideren más relevantes._

In [57]:
df_partidos.head()

Year            Datetime    Stage         Stadium         City  \
0  1930.0 1930-07-13 15:00:00  Group 1         Pocitos  Montevideo    
1  1930.0 1930-07-13 15:00:00  Group 4  Parque Central  Montevideo    
2  1930.0 1930-07-14 12:45:00  Group 2  Parque Central  Montevideo    
3  1930.0 1930-07-14 14:50:00  Group 3         Pocitos  Montevideo    
4  1930.0 1930-07-15 16:00:00  Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [58]:
df_partidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Year                  852 non-null    float64       
 1   Datetime              852 non-null    datetime64[ns]
 2   Stage                 852 non-null    object        
 3   Stadium               852 non-null    object        
 4   City                  852 non-null    object        
 5   Home Team Name        852 non-null    object        
 6   Home Team Goals       852 non-null    float64       
 7   Away Team Goals       852 non-null    float64       
 8   Away Team Name        852 non-null    object        
 9   Win conditions        852 non-null    object        
 10  Attendance            850 non-null    float64       
 11  Half-time Home Goals  852 non-null    float64       
 12  Half-time Away Goals  852 non-null    float64       
 13  Referee           

In [19]:
df_partidos.describe().transpose()

count          mean           std     min       25%  \
Year                  852.0  1.985089e+03  2.244882e+01  1930.0   1970.00   
Home Team Goals       852.0  1.811033e+00  1.610255e+00     0.0      1.00   
Away Team Goals       852.0  1.022300e+00  1.087573e+00     0.0      0.00   
Attendance            850.0  4.516480e+04  2.348525e+04  2000.0  30000.00   
Half-time Home Goals  852.0  7.089202e-01  9.374141e-01     0.0      0.00   
Half-time Away Goals  852.0  4.284038e-01  6.912519e-01     0.0      0.00   
RoundID               852.0  1.066177e+07  2.729613e+07   201.0    262.00   
MatchID               852.0  6.134687e+07  1.110572e+08    25.0   1188.75   

                          50%          75%          max  
Year                   1990.0      2002.00       2014.0  
Home Team Goals           2.0         3.00         10.0  
Away Team Goals           1.0         2.00          7.0  
Attendance            41579.5     61374.50     173850.0  
Half-time Home Goals      0.0         1.00          6.0  
Half-time Away Goals      0.0         1.00          5.0  
RoundID                 337.0    249722.00   97410600.0  
MatchID                2191.0  43950059.25  300186515.0

In [20]:
df_partidos.describe(include="object").transpose()

count unique                       top freq
Datetime             852    602      27 May 1934 - 16:30     8
Stage                852     23               Round of 16   72
Stadium              852    181            Estadio Azteca   19
City                 852    151              Mexico City    23
Home Team Name       852     78                    Brazil   82
Away Team Name       852     83                    Mexico   38
Win conditions       852     43                            787
Referee              852    366     Ravshan IRMATOV (UZB)   10
Assistant 1          852    387      ACHIK Redouane (MAR)    7
Assistant 2          852    408  KOCHKAROV Bakhadyr (KGZ)   10
Home Team Initials   852     77                       BRA   82
Away Team Initials   852     82                       MEX   38

In [27]:
years = set(df_partidos.dropna()["Year"].astype("int32"))
years

{1930,
 1934,
 1938,
 1950,
 1954,
 1958,
 1962,
 1966,
 1970,
 1974,
 1978,
 1982,
 1986,
 1990,
 1994,
 1998,
 2002,
 2006,
 2010,
 2014}

In [28]:
len(years)

20


### Descripción del dataset `df_partidos`

El dataset `df_partidos` contiene información sobre `852` partidos jugados durante `20` copas del mundo jugadas entre los años `1930` y `2014`.

El dataframe nos da tanto información *cuantitativa* como *cualitativa*. A modo de ejemplo:

- **Información cuantitativa**:
    - Goles marcados por el equipo local. En promedio se marcaron `2` goles por partido, siendo el máximo de goles marcados de `10` y el mínimo `0`
    - Goles marcados por el equipo visitante. En promedio se marcaron `1` goles por partido, siendo el máximo de goles marcados de `7` y el mínimo `0`
- **Información cualitativa**
    - Ciudad en la que se jugó el partido. Siendo *Ciudad de México* la ciudad en la que más partidos se han jugado, con un total de `23`
    - Estadio en el que se jugó el paritdo. Siendo el *Estadio Azteca* el que acumula un mayor número con `19`
    - Equipo que jugó en casa, siendo la *Selección Nacional de Brasil* la que más veces ha jugado en casa con un total de `82` veces.

## 2. Limpieza de datos

Una de las cosas que llama la atención es la alta presencia de valores perdidos en el DataFrame `df_partidos`. Cuando trabajamos con datos obtenidos del mundo real siempre nos toparemos con problemas relacionados con la medición, captura o almacenamiento de dicha información.

Localiza las filas con valores perdido. Analizar y **documentar en una nueva celda de tipo markdown** a qué se deben estos valores perdidos. Finalmente usar la función [dropna](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dropna.html) sobre el DataFrame `df_partidos` para eliminar los valores perdidos del DataFrame, y almacena el resultado en `df_partidos` de nuevo.

In [32]:
# Buscar columnas con valores nulos
df_partidos.isnull().sum()

Year                    3720
Datetime                3720
Stage                   3720
Stadium                 3720
City                    3720
Home Team Name          3720
Home Team Goals         3720
Away Team Goals         3720
Away Team Name          3720
Win conditions          3720
Attendance              3722
Half-time Home Goals    3720
Half-time Away Goals    3720
Referee                 3720
Assistant 1             3720
Assistant 2             3720
RoundID                 3720
MatchID                 3720
Home Team Initials      3720
Away Team Initials      3720
dtype: int64

In [33]:
# Analizar filas con valores nulos
df_partidos[df_partidos.Year.isnull()]

Year Datetime Stage Stadium City Home Team Name  Home Team Goals  \
852    NaN      NaN   NaN     NaN  NaN            NaN              NaN   
853    NaN      NaN   NaN     NaN  NaN            NaN              NaN   
854    NaN      NaN   NaN     NaN  NaN            NaN              NaN   
855    NaN      NaN   NaN     NaN  NaN            NaN              NaN   
856    NaN      NaN   NaN     NaN  NaN            NaN              NaN   
...    ...      ...   ...     ...  ...            ...              ...   
4567   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4568   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4569   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4570   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4571   NaN      NaN   NaN     NaN  NaN            NaN              NaN   

      Away Team Goals Away Team Name Win conditions  Attendance  \
852               NaN            NaN            NaN         NaN   
853               NaN            NaN            NaN         NaN   
854               NaN            NaN            NaN         NaN   
855               NaN            NaN            NaN         NaN   
856               NaN            NaN            NaN         NaN   
...               ...            ...            ...         ...   
4567              NaN            NaN            NaN         NaN   
4568              NaN            NaN            NaN         NaN   
4569              NaN            NaN            NaN         NaN   
4570              NaN            NaN            NaN         NaN   
4571              NaN            NaN            NaN         NaN   

      Half-time Home Goals  Half-time Away Goals Referee Assistant 1  \
852                    NaN                   NaN     NaN         NaN   
853                    NaN                   NaN     NaN         NaN   
854                    NaN                   NaN     NaN         NaN   
855                    NaN                   NaN     NaN         NaN   
856                    NaN                   NaN     NaN         NaN   
...                    ...                   ...     ...         ...   
4567                   NaN                   NaN     NaN         NaN   
4568                   NaN                   NaN     NaN         NaN   
4569                   NaN                   NaN     NaN         NaN   
4570                   NaN                   NaN     NaN         NaN   
4571                   NaN                   NaN     NaN         NaN   

     Assistant 2  RoundID  MatchID Home Team Initials Away Team Initials  
852          NaN      NaN      NaN                NaN                NaN  
853          NaN      NaN      NaN                NaN                NaN  
854          NaN      NaN      NaN                NaN                NaN  
855          NaN      NaN      NaN                NaN                NaN  
856          NaN      NaN      NaN                NaN                NaN  
...          ...      ...      ...                ...                ...  
4567         NaN      NaN      NaN                NaN                NaN  
4568         NaN      NaN      NaN                NaN                NaN  
4569         NaN      NaN      NaN                NaN                NaN  
4570         NaN      NaN      NaN                NaN                NaN  
4571         NaN      NaN      NaN                NaN                NaN  

[3720 rows x 20 columns]

In [34]:
df_partidos.loc[850:855]

Year              Datetime                     Stage  \
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   
852     NaN                   NaN                       NaN   
853     NaN                   NaN                       NaN   
854     NaN                   NaN                       NaN   
855     NaN                   NaN                       NaN   

                 Stadium             City Home Team Name  Home Team Goals  \
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   
852                  NaN              NaN            NaN              NaN   
853                  NaN              NaN            NaN              NaN   
854                  NaN              NaN            NaN              NaN   
855                  NaN              NaN            NaN              NaN   

     Away Team Goals Away Team Name                 Win conditions  \
850              3.0    Netherlands                                  
851              0.0      Argentina  Germany win after extra time    
852              NaN            NaN                            NaN   
853              NaN            NaN                            NaN   
854              NaN            NaN                            NaN   
855              NaN            NaN                            NaN   

     Attendance  Half-time Home Goals  Half-time Away Goals  \
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   
852         NaN                   NaN                   NaN   
853         NaN                   NaN                   NaN   
854         NaN                   NaN                   NaN   
855         NaN                   NaN                   NaN   

                   Referee            Assistant 1              Assistant 2  \
850  HAIMOUDI Djamel (ALG)   ACHIK Redouane (MAR)  ETCHIALI Abdelhak (ALG)   
851   Nicola RIZZOLI (ITA)  Renato FAVERANI (ITA)     Andrea STEFANI (ITA)   
852                    NaN                    NaN                      NaN   
853                    NaN                    NaN                      NaN   
854                    NaN                    NaN                      NaN   
855                    NaN                    NaN                      NaN   

      RoundID      MatchID Home Team Initials Away Team Initials  
850  255957.0  300186502.0                BRA                NED  
851  255959.0  300186501.0                GER                ARG  
852       NaN          NaN                NaN                NaN  
853       NaN          NaN                NaN                NaN  
854       NaN          NaN                NaN                NaN  
855       NaN          NaN                NaN                NaN

In [64]:
# Visualizamos el final del fichero CSV original
! tail -n 30 WorldCupMatches.csv

,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,
,,,,,,,,,,,,,,,,,,,


### Conclusiones sobre la presencia de valores nulos

Se ha observado que el dataframe contiene filas completamene vacías a partir de la línea `852` (incluida). Todo parece indicar que es debido a algún tipo de error en la generación del fichero CSV.

Dado que todos los nulos están en el final del fichero, y que los registros que contienen nulos no contienen ningún otro tipo de dato, se considera seguro usar la función `dropna()` sin que haya pérdida de información

In [35]:
df_partidos.shape

(4572, 20)

In [36]:
# Eliminación de nulos
df_partidos = df_partidos.dropna()
df_partidos.shape

(850, 20)

## 3. Cargar datos

Al igual que hemos cargado los datos de los partidos en `df_partidos`, crear las variables `df_jugadores` y `df_copas` que contengan los dataframes correspondientes a la lectura de los csv `WorldCupPlayers` y `WorldCups` (usar `pd.read_csv)`.


In [4]:
df_copas = pd.read_csv("WorldCups.csv")

In [60]:
df_jugadores = pd.read_csv("WorldCupPlayers.csv")

## 4. Rango temporal

El dataframe `df_copas` contiene datos de todos los mundiales disputados. ¿Cuál es el año del mundial más antiguo disputado? ¿Y el año del mundial más reciente?

**Pista:** recuerda que podemos usar funciones de agregación (`count` para contar el número de casos, `mean` para calcular el valor medio, etc.) directamente sobre columnas o Series de pandas. Por ejemplo, el siguiente código nos muestra el mayor número de goles marcado en un mundial.

```python
df_copas["GoalsScored"].max()
```

**Opcional:** Una vez localizados los años de interés (más antiguo y más reciente) visualizar las filas completas correspondientes a cada año usando la función `.loc`.

In [42]:
# ¿Cuál es el año del mundial más antiguo?
year_oldest = df_copas.Year.min()
year_oldest

1930

In [43]:
# ¿Cuál es el año del mundial más reciente?
year_newest = df_copas.Year.max()
year_newest

2014

In [50]:
# Visualizar las filas correspondientes a cada uno de los años
df_copas.set_index("Year").loc[[year_oldest, year_newest]]

Country   Winner Runners-Up        Third      Fourth  GoalsScored  \
Year                                                                      
1930  Uruguay  Uruguay  Argentina          USA  Yugoslavia           70   
2014   Brazil  Germany  Argentina  Netherlands      Brazil          171   

      QualifiedTeams  MatchesPlayed Attendance  
Year                                            
1930              13             18    590.549  
2014              32             64  3.386.810

In [55]:
# Visualizar las filas correspondientes a cada uno de los años (alternativa usando una mascara)
mask = (df_copas.Year == year_oldest) | (df_copas.Year == year_newest)

df_copas[mask]

Year  Country   Winner Runners-Up        Third      Fourth  GoalsScored  \
0   1930  Uruguay  Uruguay  Argentina          USA  Yugoslavia           70   
19  2014   Brazil  Germany  Argentina  Netherlands      Brazil          171   

    QualifiedTeams  MatchesPlayed Attendance  
0               13             18    590.549  
19              32             64  3.386.810